torch/_subclasses/fake_tensor.py

```diff
@@ -51,6 +52,90 @@ def _is_tensor_constructor(func: OpOverload):
 # FakeTensor extends MetaTensors to also carry an additional `fake_device`
 # which tracks devices that would have been used.
 
+def torch_dispatch_impl(cls_or_mode_instance, func, types, args, kwargs, run_function):
+    kwargs = kwargs if kwargs else {}
+
+    # This classes virtualizes .device() calls, need to short-circuit
+    # it insteead of calling device again or we would keep on recurring
+    if func == torch.ops.prim.device.default:
+        assert len(args) == 1 and isinstance(args[0], FakeTensor)
+        return args[0].fake_device
+
+    def wrap(e, device=None):
+        if isinstance(e, torch.Tensor) and not isinstance(e, FakeTensor):
+            if device:
+                return FakeTensor(e, device)
+            else:
+                return FakeTensor.from_tensor(e)
+        else:
+            return e
+
+    # if we are in the dispatch mode, we will enter this function even if the inputs
+    # are not FakeTensors. For now, throw if any non-Fake Tensor inputs
+    # and just support constructors. TODO: extend more broadly
+    if isinstance(cls_or_mode_instance, FakeTensorMode):
+        conversion_made = False
+
+        def check_non_fake_tensor(x):
+            nonlocal conversion_made
+            conversion_made = conversion_made or (isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor))
+
+        tree_map(check_non_fake_tensor, args)
+        tree_map(check_non_fake_tensor, kwargs)
```

Comment:
nit: not sure which is more idiomatic but we could do `any(tree_flatten(tree_map...)[0])` or even something like
```
conversion_made = False
for x in tree_flatten(args)[0] + tree_flatten(kwargs)[0]:
  if isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor):
    raise ...
```

this just saves us from using a nonlocal and lets us short circuit in the case that we run into the error

Reply:
IMO it's not worth optimizing for failure modes either in readability or runtime

Reply:
I'm mentally filing an issue to do `tree_iterate` separately or something along those lines because there are a lot of places in the codebase where `tree_map` is used where the return value isnt used